pyspecProcScripts/calc_error.py

```diff
@@ -0,0 +1,102 @@
+    so that the calculated variance is divided by 2.
+    By default it calculates the variance of the real of the data"""
+    assert axis is not None
+
+    def masked_var(x):
+        if var_has_imag:  # take average of variance along real and imag
+            return np.var(x[np.isfinite(x)], ddov=1) / 2
+        else:
+            return np.var(x[np.isfinite(x)], ddof=1)
+
+    return np.apply_along_axis(masked_var, axis, x)
+
+
+# }}}
+def calc_masked_error(
+    s,
+    frq_slice,
+    signal_pathway,
+    excluded_pathways=None,
+    direct="t2",
+    indirect="nScans",
+    fl=None,
+):
+    """Calculates the propagation of error for the given signal.
+
+    Before declaring the error_path,
+    look at an examples such as integration_w_error.py to see how to
+    decide which excluded pathways to take the error over.
+
+    Parameters
+    ==========
+    s: nddata
+        Full nddata set in the frequency domain for which the error is being
+        propagated.
+    frq_slice: tuple
+        Frequency range that will be filtered out in calculating the error - it
+        is assumed this region in all coherence transfer pathways contains some
+        amount of phase cycling noise.
+    signal_pathway:   dict
+                Dictionary of the path of the desired signal.
+    excluded_pathways: list
+                List of tuples containing the coherence pathways that are
+                to be masked out when calculating the error.
+                If no excluded_pathways are fed, the function will apply a mask
+                over just the signal pathway.
+    direct:     str
+                Direct axis.
+    indirect:   str
+                Indirect axis.
+    Returns
+    =======
+    collected_variance:       nddata
+             The error associated with coherence pathways not included in the
+             signal pathway.
+    """
+    # {{{ Determine if the data has an imaginary component
+    var_has_imag = False
+    if np.iscomplex(s):
+        var_has_imag = True
+    # }}}
+    phcycdims = [j for j in s.dimlabels if j.startswith("ph")]
+    collected_variance = s.C  # so we don't alter s
+    # {{{ filter out signal pathway and excluded error pathways
+    temp = select_pathway(collected_variance, signal_pathway)
+    temp.data[:] = np.nan
+    if excluded_pathways is not None:
+        error_paths = [
+            {phcycdims[i]: path[i] for i in range(len(phcycdims))}
+            for path in excluded_pathways
+        ]
+        for j in range(len(error_paths)):
+            temp = select_pathway(collected_variance, error_paths[j])
+            temp.data[:] = np.nan
+    # }}}
+    # Filter out frq_slice where ph noise resides        
+    collected_variance[direct:frq_slice].data[:] = np.nan        
+    if fl is not None:
+        fl.next("Frequency Noise")
+        fl.image(collected_variance)
+    collected_variance.run(_masked_var_multi, direct, var_has_imag)
```

Comment:
do not pass var_has_imag -- see above

Reply:
yea, so here, you want to just end with `return _masked_var_multi(collected_variance.data).item()`

Reply:
a) masked_var_multi needs an axis kwarg (the only remaining axis is indirect)
b) when I use indirect as the axis the returned value is one float
c) this would mean the error for all datapoints is the same (err = sqrt(variance*df**2*N)))
d) UNLESS you mean move the `collected_variance.run(_masked_var_multi,direct)` down to the return (so after the `_masked_mean_multi`)? but again that would not be an item

Reply:
[question](https://jmfrancklab.slack.com/archives/CV553TPRV/p1742390027557529)

Reply:
checked